netket/vqs/mc/mc_mixed_state/expect.py

```diff
@@ -0,0 +1,58 @@
+# Copyright 2021 The NetKet Authors - All rights reserved.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#    http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+from functools import partial
+from typing import Callable
+
+import numpy as np
+
+from netket.utils.dispatch import dispatch
+
+from netket.operator import (
+    DiscreteOperator,
+    AbstractSuperOperator,
+    Squared,
+)
+
+from netket.vqs.mc import kernels, check_hilbert, get_configs, get_fun
+
+from .state import MCMixedState
+
+
+@dispatch
+def get_configs(vstate: MCMixedState, Ô: DiscreteOperator):
+    check_hilbert(vstate.diagonal.hilbert, Ô.hilbert)
+
+    σ = vstate.diagonal.samples
+    σp, mels = Ô.get_conn_padded(σ)
+    return σ, σp, mels
+
+
+@dispatch
+def get_fun(vstate: MCMixedState, Ô: DiscreteOperator):
+    return kernels.local_value_op_op_cost
+
+
+@dispatch
+def get_configs(vstate: MCMixedState, Ô: Squared[AbstractSuperOperator]):
```

Comment:
what is the different between get_configs, get_conn, get_conn_padded etc? why do we need to dispatch over get_configs?

Reply:
I still need to write some docs about this, but this should be quite clear from the two dispatch rules in this file.

This function is basically equivalent to `_pack_data` of @gpescia's PR. Eventually I want to merge them.
It prepares some data to be passed to `jax.jit` depending on both the variational state type and the operator type.

For standard operators and MCState, this is just `get_conn_padded`.
For ~batched~ chunked operators this is `get_conn_flattened`.

For the expectation value of an operator over a mixed state, this needs to take the samples of the diagonal.
For the expectation value of LdagL we need to take the connected elements of L and not of the whole LdagL because the algorithm is written differently.

If we write a correlation operator <AB>-<A><B>, we'd need to call get conn twice, on both A and B...

The dispatch is there to select which logic we need

Reply:
ok but the abstract dispatch logic on get_configs seems tied to the lattice implementation since it returns  a tuple with 3 elements (sigma, sigmap, mels)

Reply:
Indeed.
This is what I am missing to actually unify the two interfaces: do the same thing as in continuous and pack this stuff into a pytree/tuple.

Then they are basically the same interface

Reply:
OK but does this mean that the dispatch logic in `expect` can be later removed in favour of this dispatched kernel+data approach? If this is the case, then I agree this is going to simplify the logic... I was just worried that we would keep the dispatch logic also on `expect` (and `expect_and_grad`)

Reply:
to make life easier for users wishing to define their custom operators, we cannot.

In a sense, before this PR one had to write a function for every new operator defined to compute expect.
With this PR there is a general high level function that should work for most cases (_expect_hermitian) and you only have to define (through dispatch) the inner kernel computing the local-estimator and a function to prepare some data for it (sigmap, mels...).

This is an improvement, because most cases will be covered by the high level rule and therefore you only need to define the local estimator.

But this cannot cover everything. So if we still want to allow users to do *anything*, we need to leave a possibility for an user to completely override the logic we have written.

Reply:
ok let's discuss later, I think we can remove one level of dispatch but maybe I am wrong

Reply:
such an example is given by the implementations for `exactState`. They cannot be written (efficiently) using `_expect_hermitian`. They need to use a different macro-kernel.

Reply:
yeah, let's discuss offline